Src/WindowsServer/WindowsServer.Shared/FirstChanceExceptionStatisticsTelemetryModule.cs

```diff
@@ -248,21 +321,100 @@ private void TrackStatistics(string type, string operation, string method, bool
             }
         }
 
-        private string GetDimCappedString(string dimensionValue, ConcurrentDictionary<string, int> capValues)
+        private string GetDimCappedString(string dimensionValue, HashCache<string> hashCache)
         {
-            int temp;
-            if (capValues.TryGetValue(dimensionValue, out temp))
+            hashCache.RwLock.EnterReadLock();
+
+            if (hashCache.ValueCache.Contains(dimensionValue) == true)
             {
+                hashCache.RwLock.ExitReadLock();
+
                 return dimensionValue;
             }
-            else if (capValues.Count > 100)
+
+            if (hashCache.ValueCache.Count > CacheSize)
             {
+                hashCache.RwLock.ExitReadLock();
+
                 return "OtherValue";
             }
-            else
+
+            hashCache.RwLock.ExitReadLock();
+            hashCache.RwLock.EnterWriteLock();
+
+            hashCache.ValueCache.Add(dimensionValue);
+
+            hashCache.RwLock.ExitWriteLock();
+
+            return dimensionValue;
+        }
+
+        private void ResetDimCapCaches(HashCache<string> cache1, HashCache<string> cache2, HashCache<string> cache3)
+        {
+            cache1.RwLock.EnterWriteLock();
+
+            if (this.DimCapTimeout < DateTime.UtcNow.Ticks)
             {
-                capValues.TryAdd(dimensionValue, 0);
-                return dimensionValue;
+                cache2.RwLock.EnterWriteLock();
+
+                if (this.DimCapTimeout < DateTime.UtcNow.Ticks)
+                {
+                    cache3.RwLock.EnterWriteLock();
+
+                    if (this.DimCapTimeout < DateTime.UtcNow.Ticks)
+                    {
+                        cache1.ValueCache.Clear();
+                        cache2.ValueCache.Clear();
+                        cache3.ValueCache.Clear();
+
+                        this.DimCapTimeout = DateTime.UtcNow.Ticks + this.cacheLifetime;
+                    }
+
+                    cache3.RwLock.ExitWriteLock();
+                }
+
+                cache2.RwLock.ExitWriteLock();
+            }
+
+            cache1.RwLock.ExitWriteLock();
+        }
+
+        internal class HashCache<T> : IDisposable
```

Comment:
From my testing, this is less overhead and more performant than using Concurrent Dictionary

---
In reply to: [107244751](https://github.com/Microsoft/ApplicationInsights-dotnet-server/pull/370#discussion_r107244751) [](ancestors = 107244751)